Makefile

```diff
@@ -188,22 +189,37 @@ ko-build-transport-local: ## Build local transport container image with `ko`.
 
 # this is used for local testing
 .PHONY: kind-load-image
-kind-load-image:
+kind-load-image: ko-build-controller-manager-local ko-build-transport-local
 	kind load --name ${KIND_HOSTING_CLUSTER} docker-image ${CONTROLLER_MANAGER_IMAGE}
+	kind load --name ${KIND_HOSTING_CLUSTER} docker-image ${TRANSPORT_IMAGE}
 
 # this is used for local testing
 .PHONY: k3d-load-image
-k3d-load-image:
+k3d-load-image: ko-build-controller-manager-local ko-build-transport-local
 	k3d image import ${CONTROLLER_MANAGER_IMAGE} -c ${K3D_HOSTING_CLUSTER}
+	k3d image import ${CONTROLLER_MANAGER_IMAGE} -c ${TRANSPORT_IMAGE}
 
-.PHONY: chart
-chart: manifests kustomize
-	cd config/manager && $(KUSTOMIZE) edit set image controller=$(shell echo ${CONTROLLER_MANAGER_IMAGE} | sed 's/\(:.*\)v/\1/')
+.PHONY: core-chart
+core-chart: manifests kustomize
 	$(KUSTOMIZE) build config/default \
-		| yq '. | select(.kind == "ClusterRole*").metadata.name |= "{{.Values.ControlPlaneName}}-" + .' \
-		| yq '. | select(.kind == "ClusterRoleBinding").roleRef.name |= "{{.Values.ControlPlaneName}}-" + .' \
-		> chart/templates/controller-manager.yaml
+		| yq '. | select(.kind == "ClusterRole*").metadata.name |= "{{\"{{.ControlPlaneName}}\"}}-" + .' \
+		| yq '. | select(.kind == "ClusterRoleBinding").roleRef.name |= "{{\"{{.ControlPlaneName}}\"}}-" + .' \
+		| yq '[.]' \
+		| sed -e 's/^/  /' \
+		> /tmp/kubestellar-controller-manager.yaml
 	git checkout -- config/manager/kustomization.yaml
+	sed -i -e '/KUBESTELLAR_CONTROLLER_PLACEHOLDER/{r /tmp/kubestellar-controller-manager.yaml' -e 'd}' core-chart/templates/postcreatehooks/wds.yaml
```

Comment:
This is not working for me.

```console
mspreitz@mjs13 kubestellar % make core-chart
test -s /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/bin/controller-gen && /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/bin/controller-gen --version | grep -q v0.13.0 || \
	GOTOOLCHAIN=go1.21.8 GOBIN=/Users/mspreitz/go/src/github.com/kubestellar/kubestellar/bin go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.13.0
test -s /Users/mspreitz/go/src/github.com/kubestellar/kubestellar/bin/kustomize || GOBIN=/Users/mspreitz/go/src/github.com/kubestellar/kubestellar/bin go install sigs.k8s.io/kustomize/kustomize/v5@v5.3.0
/Users/mspreitz/go/src/github.com/kubestellar/kubestellar/bin/controller-gen rbac:roleName=manager-role crd webhook paths="./api/..." output:crd:artifacts:config=config/crd/bases
/Users/mspreitz/go/src/github.com/kubestellar/kubestellar/bin/kustomize build config/crd/ > pkg/crd/files/crds.yaml
/Users/mspreitz/go/src/github.com/kubestellar/kubestellar/bin/kustomize build config/default \
		| yq '. | select(.kind == "ClusterRole*").metadata.name |= "{{\"{{.ControlPlaneName}}\"}}-" + .' \
		| yq '. | select(.kind == "ClusterRoleBinding").roleRef.name |= "{{\"{{.ControlPlaneName}}\"}}-" + .' \
		| yq '[.]' \
		| sed -e 's/^/  /' \
		> /tmp/kubestellar-controller-manager.yaml
# Warning: 'patchesStrategicMerge' is deprecated. Please use 'patches' instead. Run 'kustomize edit fix' to update your Kustomization automatically.
git checkout -- config/manager/kustomization.yaml
sed -i -e '/KUBESTELLAR_CONTROLLER_PLACEHOLDER/{r /tmp/kubestellar-controller-manager.yaml' -e 'd}' core-chart/templates/postcreatehooks/wds.yaml
sed: 1: "/KUBESTELLAR_CONTROLLER ...": unexpected EOF (pending }'s)
make: *** [core-chart] Error 1
```

Reply:
Also, stuffing a whole file unescaped into the `replacement` part of a `s/regexp/replacement/flags` command is problematic, because the `replacement` is interpreted.

Reply:
I'm not sure what the problem is. It's working for me.

```shell
$ sed --version
sed (GNU sed) 4.8
```

```shell
$ make core-chart
test -s /vagrant/kubestellar/bin/controller-gen && /vagrant/kubestellar/bin/controller-gen --version | grep -q v0.13.0 || \
GOTOOLCHAIN=go1.21.8 GOBIN=/vagrant/kubestellar/bin go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.13.0   
test -s /vagrant/kubestellar/bin/kustomize || GOBIN=/vagrant/kubestellar/bin go install sigs.k8s.io/kustomize/kustomize/v5@v5.3.0
/vagrant/kubestellar/bin/controller-gen rbac:roleName=manager-role crd webhook paths="./api/..." output:crd:artifacts:config=config/crd/bases
/vagrant/kubestellar/bin/kustomize build config/crd/ > pkg/crd/files/crds.yaml
/vagrant/kubestellar/bin/kustomize build config/default \
        | yq '. | select(.kind == "ClusterRole*").metadata.name |= "{{\"{{.ControlPlaneName}}\"}}-" + .' \
        | yq '. | select(.kind == "ClusterRoleBinding").roleRef.name |= "{{\"{{.ControlPlaneName}}\"}}-" + .' \
        | yq '[.]' \
        | sed -e 's/^/  /' \
        > /tmp/kubestellar-controller-manager.yaml
# Warning: 'patchesStrategicMerge' is deprecated. Please use 'patches' instead. Run 'kustomize edit fix' to update your Kustomization automatically.
git checkout -- config/manager/kustomization.yaml
sed -i -e '/KUBESTELLAR_CONTROLLER_PLACEHOLDER/{r /tmp/kubestellar-controller-manager.yaml' -e 'd}' core-chart/templates/postcreatehooks/wds.yaml
```

Reply:
> Also, stuffing a whole file unescaped into the `replacement` part of a `s/regexp/replacement/flags` command is problematic, because the `replacement` is interpreted.

Perhaps but it's working. All will be removed once we remove the kustomization,
I'm really trying to do this in 2 steps instead one one.

Reply:
FYI:

```console
mspreitz@mjs13 kubestellar % fgrep -n '# KUBESTELLAR_CONTROLLER_PLACEHOLDER' core-chart/templates/postcreatehooks/wds.yaml              
166:  # KUBESTELLAR_CONTROLLER_PLACEHOLDER
mspreitz@mjs13 kubestellar % 
mspreitz@mjs13 kubestellar % fgrep -n '# KUBESTELLAR_CONTROLLER_PLACEHOLDER' core-chart/templates/postcreatehooks/wds.yaml | cut -d: -f1
166
```

Reply:
```shell
$ fgrep -n '# KUBESTELLAR_CONTROLLER_PLACEHOLDER' core-chart/templates/postcreatehooks/wds.yaml
166:  # KUBESTELLAR_CONTROLLER_PLACEHOLDER

$ make core-chart
test -s /vagrant/kubestellar/bin/controller-gen && /vagrant/kubestellar/bin/controller-gen --version | grep -q v0.13.0 || \
GOTOOLCHAIN=go1.21.8 GOBIN=/vagrant/kubestellar/bin go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.13.0
test -s /vagrant/kubestellar/bin/kustomize || GOBIN=/vagrant/kubestellar/bin go 
install sigs.k8s.io/kustomize/kustomize/v5@v5.3.0
/vagrant/kubestellar/bin/controller-gen rbac:roleName=manager-role crd webhook paths="./api/..." output:crd:artifacts:config=config/crd/bases
/vagrant/kubestellar/bin/kustomize build config/crd/ > pkg/crd/files/crds.yaml
/vagrant/kubestellar/bin/kustomize build config/default \
        | yq '. | select(.kind == "ClusterRole*").metadata.name |= "{{\"{{.ControlPlaneName}}\"}}-" + .' \
        | yq '. | select(.kind == "ClusterRoleBinding").roleRef.name |= "{{\"{{.ControlPlaneName}}\"}}-" + .' \
        | yq '[.]' \
        | sed -e 's/^/  /' \
        > /tmp/kubestellar-controller-manager.yaml
# Warning: 'patchesStrategicMerge' is deprecated. Please use 'patches' instead. 
Run 'kustomize edit fix' to update your Kustomization automatically.
git checkout -- config/manager/kustomization.yaml
sed -i -e '/KUBESTELLAR_CONTROLLER_PLACEHOLDER/{r /tmp/kubestellar-controller-manager.yaml' -e 'd}' core-chart/templates/postcreatehooks/wds.yaml

$ fgrep -n '# KUBESTELLAR_CONTROLLER_PLACEHOLDER' core-chart/templates/postcreatehooks/wds.yaml
$

```

Reply:
FYI, I find that the following works. Except for the fact that it injects LF-terminated lines into a file that is CRLF-terminated.

```
	(where=$$(fgrep -n '# KUBESTELLAR_CONTROLLER_PLACEHOLDER' core-chart/templates/postcreatehooks/wds.yaml | cut -d: -f1) &&\
	echo where=[$$where] &&\
	head -n $$(( where-1 )) core-chart/templates/postcreatehooks/wds.yaml > replaced.yaml &&\
	cat /tmp/kubestellar-controller-manager.yaml >> replaced.yaml &&\
	tail -n +$$(( where+1 )) core-chart/templates/postcreatehooks/wds.yaml >> replaced.yaml )
	mv replaced.yaml core-chart/templates/postcreatehooks/wds.yaml
```

Reply:
See https://github.com/MikeSpreitzer/kcp-edge-mc/commits/merge-charts/

Reply:
It needs to work for developers using a Mac.
I am using MacOS Sonoma, make version 3.81, my bash version is 3.2.57, and my default shell is zsh version 5.9.

Reply:
Oh, I misread the failing command. It is not a replace command, it is using a pattern to identify a line and then applied a pair of `r` and `d` commands to that line. I suspect the problem is just in the way that the bracketed pair of `r` and `d` commands is formed is the problem.

Reply:
I am getting closer with the following.

```
	sed -e $$'/KUBESTELLAR_CONTROLLER_PLACEHOLDER/{=; r /tmp/kubestellar-controller-manager.yaml ; d\n}' -i.bak core-chart/templates/postcreatehooks/wds.yaml
```

`man sed` says that the closing curly bracket has to be preceded by a newline. But the replacement is not happening.

Reply:
OK, I am succeeding with the following.

```
	sed -e $$'/KUBESTELLAR_CONTROLLER_PLACEHOLDER/{r /tmp/kubestellar-controller-manager.yaml\n d\n}' -i.bak core-chart/templates/postcreatehooks/wds.yaml
```

Except that:
1. This still mixes LF-terminated content into a CRLF-terminated file.
2. This creates a `.bak` file. MacOS and Linux are ferociously incompatible in how to tell `sed` to process a file in place without creating a backup.

Reply:
If you can provide code that works on both Linux and Mac, I personally don't care.
This will all be removed soon after the PR is merged

Reply:
OK, I think I have something that works on both Mac and Linux. I tested it on MacOS Sonoma and Ubuntu 22.04.

Rather than go to all the lengths necessary to cope with the `-i` divergence between `sed` on Linux and MacOS, I took a different approach --- distinct input and output files. This has the added advantage that `make core-chart` no longer is modifying a file that is under source code control.

See https://github.com/MikeSpreitzer/kcp-edge-mc/tree/merge-charts

Reply:
@francostellari has taken a similar but different approach, using `sed -i.bak ..; rm ....bak`. I would prefer to not be in the business of modifying a file under Git control. This can be pursued in follow-on work.